src/jquery.multiselect.filter.js

```diff
@@ -97,98 +124,142 @@
 
-      // reference to the actual inputs
-      this.$inputs = this.instance.$inputs;
+      // If menu already opened, have to reset menu height since
+      // addition of the filter input changes the header height calc.
+      if (!!this.instance._isOpen) {
+         this.instance._setMenuHeight(true);
+      }
 
       // cache input values for searching
       this.updateCache();
 
       // Change the normal _toggleChecked fxn behavior so that when checkAll/uncheckAll
-      // is fired, only the currently displayed filtered inputs are checked
-      var $instance = this.instance;
-      $instance._oldToggleChecked = $instance._toggleChecked;
-      $instance._toggleChecked = function(flag, group) {
-         $instance._oldToggleChecked(flag, group, true);
+      // is fired, only the currently displayed filtered inputs are checked if filter entered.
+      var instance = this.instance,
+            filter = this.$input[0];
+      instance._oldToggleChecked = instance._toggleChecked;
+      instance._toggleChecked = function(flag, group) {
+         instance._oldToggleChecked(flag, group, !!filter.value);
       };
 
     },
 
-    // thx for the logic here ben alman
+   /**
+    * Handles searches as text is entered in the filter box.
+    * Uses a text cache to speed up searching.
+    * Debouncing is done to limit how often this is ran.
+    * Alternate filter rules can be used.
+    * Option group labels may be searched, also.
+    * @param (object) event object from original event.
+    */
     _handler: function(e) {
-      var term = $.trim(this.$input[0].value.toLowerCase()),
-
-      // speed up lookups
-      $rows = this.$rows, $inputs = this.$inputs, $cache = this.$cache;
-      var $groups = this.instance.$menu.find(".ui-multiselect-optgroup");
-      $groups.show();
-      $rows.toggle(!term);
-      if(term) {
-        var regex = new RegExp(term.replace(rEscape, "\\$&"), 'gi');
-
-        this._trigger("filter", e, $.map($cache, function(v, i) {
-          if(v.search(regex) !== -1) {
-            $rows.eq(i).show();
-            return $inputs.get(i);
+      var term = this.$input[0].value.toLowerCase().replace(/^\s+|\s+$/g,''),
+            filterRule = this.options.filterRule || 'contains',
+            regex = new RegExp( ( filterRules[filterRule] || filterRule ).replace('{{term}}', term.replace(rEscape, "\\$&")), 'i'),
+            searchGroups = !!this.options.searchGroups,
+            $checkboxes = this.instance.$checkboxes,
+            cache = this.cache,   // Cached text() object
+            optgroupClass = "ui-multiselect-optgroup",
+            hiddenClass = 'ui-multiselect-excluded';
+
+      this.$rows.toggleClass(hiddenClass, !!term);
+      if (!searchGroups) {
+         // If not searching in groups then show all group headings in the results.
+         $checkboxes.find('.' + optgroupClass).removeClass(hiddenClass);
+      }
+      var filteredInputs = $checkboxes.children().map(function(x) {
```

Comment:
These are just index variables...  what would you suggest as an alternative... index1 and index2?

Reply:
`groupIndex` and `optionIndex`?

Reply:
@mlh758 However, if there are no groups, then groupIndex would actually be the optionIndex.  
(There would only be options.)   Thus, this would be confusing for the no groups case.

Reply:
Yeah that wouldn't be any good. We could probably reassess what it looks like after the filtering issue is worked out.

Reply:
Yeah, leave it as `x` and `y` for now then.